executor/aggregate.go

```diff
@@ -698,54 +879,91 @@ func (e *HashAggExec) execute(ctx context.Context) (err error) {
 		if e.childResult.NumRows() == 0 {
 			return nil
 		}
+
+		err = e.getGroupKey(e.childResult)
+		if err != nil {
+			return err
+		}
+
+		partialResults := e.getPartialResults(e.groupKeys)
 		for row := inputIter.Begin(); row != inputIter.End(); row = inputIter.Next() {
-			groupKey, err := e.getGroupKey(row)
-			if err != nil {
-				return err
-			}
-			if !e.groupSet.Exist(groupKey) {
-				e.groupSet.Insert(groupKey)
-				e.groupKeys = append(e.groupKeys, groupKey)
-			}
-			partialResults := e.getPartialResults(groupKey)
+			j := 0
 			for i, af := range e.PartialAggFuncs {
-				err = af.UpdatePartialResult(e.ctx, []chunk.Row{row}, partialResults[i])
+				err = af.UpdatePartialResult(e.ctx, []chunk.Row{row}, partialResults[j][i])
 				if err != nil {
 					return err
 				}
 			}
+			j++
 		}
 	}
 }
 
-func (e *HashAggExec) getGroupKey(row chunk.Row) (string, error) {
-	e.groupValDatums = e.groupValDatums[:0]
+func (e *HashAggExec) getGroupKey(input *chunk.Chunk) (err error) {
+	numRows := input.NumRows()
+	if numRows > len(e.groupKeyBuffer) {
+		for i := 0; i < len(e.groupKeyBuffer); i++ {
+			deallocateByteSlice(e.groupKeyBuffer[i])
+		}
+		deallocateGroupKeySlice(e.groupKeyBuffer)
+		e.groupKeyBuffer = allocGroupKeySlice(numRows)
+		e.groupKeyBuffer = e.groupKeyBuffer[:0]
+		for i := 0; i < numRows; i++ {
+			e.groupKeyBuffer = append(e.groupKeyBuffer, allocByteSlice(len(e.groupKeyBuffer)))
+		}
+	} else {
+		for i := numRows; i < len(e.groupKeyBuffer); i++ {
+			deallocateByteSlice(e.groupKeyBuffer[i])
+		}
+		e.groupKeyBuffer = e.groupKeyBuffer[:numRows]
+	}
```

Comment:
How about：
	for len(e.groupKeyBuffer) < numRows {
		e.groupKeyBuffer = append(e.groupKeyBuffer, make([]byte, 0, 256))
	}